src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -0,0 +1,978 @@
+        if (UNLIKELY(gpio_mode == GPIO_INVALID_MODE)) {
+            AVM_LOGE(TAG, "Invalid gpio mode: %s", interop_atom_to_string(ctx, mode_atom));
+            return ERROR_ATOM;
+        }
+        if ((gpio_mode == GPIO_MODE_OUTPUT) || (gpio_mode == GPIO_MODE_OUTPUT_OD)) {
+            if (gpio_mode == GPIO_MODE_OUTPUT_OD) {
+                gpio_mode = GPIO_MODE_OUTPUT;
+                out_type = GPIO_OTYPE_OD;
+                setup_output = true;
+            } else {
+                out_type = GPIO_OTYPE_PP;
+                setup_output = true;
+            }
+        }
+        pull_atom = term_get_tuple_element(mode_term, 1);
+        if (UNLIKELY(!term_is_atom(pull_atom))) {
+            AVM_LOGE(TAG, "GPIO pull direction must be an atom ('up', 'down', or 'floating').");
+            return ERROR_ATOM;
+        }
+        pull_up_down = (uint8_t) interop_atom_term_select_int(pull_mode_table, pull_atom, ctx->global);
+        if ((setup_output) && (term_get_tuple_arity(mode_term) == 3)) {
+            mhz_atom = term_get_tuple_element(mode_term, 2);
+            if (UNLIKELY(!term_is_atom(mhz_atom))) {
+                AVM_LOGE(TAG, "GPIO output speed must be an atom ('mhz_2', 'mhz_25', 'mhz_50', or 'mhz_100').");
+                return ERROR_ATOM;
+            }
+            output_speed = (uint8_t) interop_atom_term_select_int(output_mhz_table, mhz_atom, ctx->global);
+            if (output_speed == INVALID_GPIO_OSPEED) {
+                output_speed = GPIO_OSPEED_2MHZ;
+                AVM_LOGW(TAG, "Invalid output speed given, falling back to 2 Mhz default.");
+            }
+        } else if (setup_output) {
+            output_speed = GPIO_OSPEED_2MHZ;
+            AVM_LOGW(TAG, "No output speed given, falling back to 2 Mhz default.");
+        }
+    } else {
+        mode_atom = mode_term;
+        if (UNLIKELY(!term_is_atom(mode_atom))) {
+            AVM_LOGE(TAG, "GPIO Mode must be an atom ('input', 'output', 'output_od').");
+            return ERROR_ATOM;
+        }
+        gpio_mode = (uint8_t) interop_atom_term_select_int(pin_mode_table, mode_atom, ctx->global);
+        if (UNLIKELY(gpio_mode == GPIO_INVALID_MODE)) {
+            AVM_LOGE(TAG, "Invalid gpio mode: %s", interop_atom_to_string(ctx, mode_atom));
+            return ERROR_ATOM;
+        }
+        pull_up_down = GPIO_PUPD_NONE;
+        if (setup_output) {
+            output_speed = GPIO_OSPEED_2MHZ;
+        }
+    }
+
+    gpio_mode_setup(gpio_bank, gpio_mode, pull_up_down, gpio_pin_mask);
+    if (setup_output) {
+        gpio_set_output_options(gpio_bank, out_type, output_speed, gpio_pin_mask);
+        AVM_LOGD(TAG, "Setup: Pin bank '%s' pin bitmask 0x%04X %s mode (0x%02X), output speed %s, pull mode %s (0x%02X)", interop_atom_to_string(ctx, gpio_bank_atom), gpio_pin_mask, interop_atom_to_string(ctx, mode_atom), gpio_mode, interop_atom_to_string(ctx, mhz_atom), interop_atom_to_string(ctx, pull_atom), pull_up_down);
```

Comment:
leaks

(I won’t catch all of them below, so just search for `interop_atom_to_string`)